### ISSUE TITLE
Strategies for Queryable structs that need a dynamic value in SwiftUI

### ISSUE THREAD

**Issue by DreamingInBinary** [groue/GRDBQuery#11]
In the demo application dealing with the `Player` model, consider if you had a model called `Team`, which has many `Player` identifiers associated to it. 

In such a case, you might have a view where you can toggle the `Team` you're viewing. To get at the data, I'd construct such a `Query` type like this to drive it:

```swift
struct TeamQueryableRequest: Queryable {
    var teamId: Int64
    static var defaultValue: [Team] { [] }
    
    func publisher(in appDatabase: AppDatabase) -> AnyPublisher<[Team], Error> {
        ValueObservation
            .tracking(fetchValue(_:))
            .publisher(
                in: appDatabase.databaseReader,
                scheduling: .immediate)
            .eraseToAnyPublisher()
    }
    
    func fetchValue(_ db: Database) throws -> [Release] {
        return try Team(id: teamId).fetchAll(db)
    }
}
``` 

If you just present the `View` to see a `Team`  in a one-off manner such as a modal sheet, such as in the demo application, then it is acceptable to pass the `id` in the initializer:

```swift
struct TeamView: View {    
    @Query< TeamQueryableRequest >
    private var teams: [Team]
        
    init(id: Int64) {
        _ teams = Query(TeamQueryableRequest(teamId: id))
    }
}
```

The issue I am having is this: That view is always displayed (this is a Mac app I'm dealing with, where the view hierarchy is much more flattened than iOS) and the `id` changes dynamically as the user toggles which `Team` they are viewing from an `EnvironmentObject` that is passed around, i.e.:

```swift
class Store: ObservableObject {
    @Published var selectedTeamId: Int64 = 0
}
```

Due to Swift's strict initializer rules, I'm unable to do something like this:

```swift
struct TeamView: View {    
    @EnvironmentObject var store: Store
    @Query< TeamQueryableRequest >
    private var teams: [Team]
       
    init(id: Int64) {
        _ teams = Query(TeamQueryableRequest(teamId: store.selectedTeamId)) ~~~Compiler Error~~~
    }
}
```

**TL;DR:** How do I leverage `Query` when it relies on data that can't be used until after initialization? 

Is there a good way to handle this, it feels like I'm making it too difficult. Thank you for any advice, and happy to expand if I need to.

**Comment by groue** [comment]
Hello @DreamingInBinary,

I technique I know is to nest the `@Query` inside another view in your `TeamView`, as in the sample code below:

```swift
struct TeamView: View {    
    @EnvironmentObject var store: Store
       
    var body: some View {
        NestedTeamView(teamId: store.selectedTeamId)
    }
}

struct NestedTeamView: View {    
    @Query<TeamQueryableRequest> private var teams: [Team]
       
    init(teamId: Int64) {
        _ teams = Query(TeamQueryableRequest(teamId: teamId))
    }
}
```

**Comment by DreamingInBinary** [comment]
@groue thanks, these are great ways to look at the problem. I went with your first solution. I love the idea of `QueryView`, but I also know exactly what you mean with the API layering. SwiftUI is very abstract to begin with, so I'm trying to maintain as much control over it as I can. But, if that does make it to your first major release in this repo, I'll definitely give it a spin.

I can close this issue if you want - not sure if you prefer to have them closed out or to remain open.

**Comment by groue** [comment]
If your problem is solved, we can close the issue 🙂
